### PR TITLE
RFC: Remove circular dependency between fetcher(uptane) and packagemanager

### PR DESCRIPTION
**Pull request by doanac**
This is a very lightly tested attempt at cleaning up the relationship between the fetcher, sotauptaneclient, and packagemanager.

Its dependent on PR #1122. Its kind of a prereq issue #1118 so that the "app download" logic won't make the existing fetcher logic even messier.

### CODE REVIEW

File: src/libaktualizr/package_manager/fetcher_test.cc [hunk 1]
```diff
@@ -56,7 +64,10 @@ void test_pause(const Uptane::Target& target) {
 
   std::shared_ptr<INvStorage> storage(new SQLStorage(config.storage, false));
   auto http = std::make_shared<HttpClient>();
-  Uptane::Fetcher f(config, storage, http, progress_cb);
+
+  auto pacman = std::make_shared<PKGMGR_TYPE>(config.pacman, storage, nullptr, http);
```
Comment (lbonn) [inline src/libaktualizr/package_manager/fetcher_test.cc]: We try to avoid having tests that run differently depending on CMake flags, so that we can be confident we test everything when the test suite runs with everything `=on`. The CMake flags are only here to add functionality, not to configure it.

In this case, `test_pause()` could have the package manager type as a template parameter, or could take a `std::function` factory.

File: src/libaktualizr/package_manager/ostreemanager.cc [hunk 2]
```diff
@@ -222,14 +222,22 @@ data::InstallationResult OstreeManager::finalizeInstall(const Uptane::Target &ta
...
   if (sysroot_smart == nullptr) {
     throw std::runtime_error("Could not find OSTree sysroot at: " + config.sysroot.string());
   }
 }
 
+bool OstreeManager::fetchTarget(const Uptane::Target &target, const std::string &repo_server, const KeyManager &keys,
+                                FetcherProgressCb progress_cb, const api::FlowControlToken *token) {
+  if (!target.IsOstree()) {
+    return PackageManagerBase::fetchTarget(target, repo_server, keys, progress_cb, token);
+  }
```
Comment (patrickvacek) [inline src/libaktualizr/package_manager/ostreemanager.cc]: I was going to say this should be an error (installing a binary file on an OSTree filesystem) but apparently our code allows it at present. I'm not sure that's right, but that's also not something you should worry about for now unless you are depending upon that logic.

File: src/libaktualizr/package_manager/fetcher_test.cc (hunk 1 above)
Comment (doanac) [reply src/libaktualizr/package_manager/fetcher_test.cc]: good idea. thanks.

File: src/libaktualizr/package_manager/ostreemanager.cc (hunk 2 above)
Comment (doanac) [reply src/libaktualizr/package_manager/ostreemanager.cc]: Correct. I was really trying to make this function *exactly* the same as the old code. However, I think that even though the download passes in the current code base, the install step will fail. So its not a huge difference and something that would be nice to clean up at some point.

It would also probably be smart to update the Web UI to not lead users to think they can install binary files on an OSTree based device.

TBH - I'm a bit confused what the point of installing a file is at all. I guess maybe its applicable to the DebianPackageManager?

That also got me wondering about the DebianPackageManager. Is this really used by anyone? It looks like its only capable of installing a single package, so I was confused how someone might use that?

**Comment by doanac** [comment]
>What is the point of distinguishing between PackageManagerBase and PackageManagerInterface? 

I could easily merge this. My first attempt did, and then I split it out. I prefer the simplicity of just having one class.

> I recommend in any event waiting for #1122 to get merged first and then rebasing this on master.

fantastic. I'll table this for now and we can pick this back up when its merged. FYI - you may want to get that PR rebased on master. I think it may have a merge conflict which would be nice to resolve while its still being tested.

File: src/libaktualizr/package_manager/ostreemanager.cc (hunk 2 above)
Comment (patrickvacek) [reply src/libaktualizr/package_manager/ostreemanager.cc]: > However, I think that even though the download passes in the current code base, the install step will fail. So its not a huge difference and something that would be nice to clean up at some point.

That matches my expectations. Not a high priority right now, but I agree we should clean that up (and on the web UI if possible, too).

> I'm a bit confused what the point of installing a file is at all.

There are at least three uses that I am aware of at present:
1. Non-OSTree primaries that use an external installer. This comes up in customer projects. However, all of the examples of that are currently closed-source.
2. Secondaries. Most examples we've seen do not use OSTree and just flash a whole filesystem image.
3. The fake package manager, which is great for testing.

> That also got me wondering about the DebianPackageManager. Is this really used by anyone?

In short, no. It was a PoC that we got working at a basic level and then it never went anywhere. It is mostly dead weight, but it is kind of interesting, so we've kept it around in case it ever becomes useful again.

**Comment by patrickvacek** [comment]
> FYI - you may want to get that PR rebased on master. I think it may have a merge conflict which would be nice to resolve while its still being tested.

You are probably correct. @eu-smirnov is still working on that PR and may already be aware of the conflict, but now he's been tagged and informed if he wasn't already!

**Comment by patrickvacek** [comment]
@doanac The new PR looks very promising, and if this is just the draft version, we can close this now, right?
